README.md
[![Build Status](https://travis-ci.org/zerointermittency/redis.svg?branch=master)](https://travis-ci.org/zerointermittency/redis)
[![Maintainability](https://api.codeclimate.com/v1/badges/c3afc1fac7199fbbc9d5/maintainability)](https://codeclimate.com/github/zerointermittency/redis/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/c3afc1fac7199fbbc9d5/test_coverage)](https://codeclimate.com/github/zerointermittency/redis/test_coverage)
[![npm version](https://badge.fury.io/js/%40zerointermittency%2Fredis.svg)](https://badge.fury.io/js/%40zerointermittency%2Fredis)
[![Downloads](https://img.shields.io/npm/dt/@zerointermittency/redis.svg)](https://www.npmjs.com/package/@zerointermittency/redis)
[![dependency status](https://david-dm.org/zerointermittency/redis.svg)](https://david-dm.org/zerointermittency/redis)
[![devDependency status](https://david-dm.org/zerointermittency/redis/dev-status.svg)](https://david-dm.org/zerointermittency/redis)

# Bienvenido

Este modulo pretende tener la estrategia para manejar el cache en **[redis][redis]** de todos los módulos y proyectos.

## Instalación

```bash
yarn add @zerointermittency/redis
# npm i --save @zerointermittency/redis
```

## Errores estandarizados

Código | Nombre | Mensaje
-------|--------|----------------------------
100    |type    |"{varName}" must be "{type}"
101    |empty   |Empty var {varName}

## Api

El modulo extiende de la clase del modulo **[ioredis][ioredis]**, para poder hacer algunas validaciones, conservar todos los métodos que nos provee **[ioredis][ioredis]** y mejorar la estrategia de redis que el ya posee.

##### Iniciar

Se debe instanciar un objeto como se hace a continuación:

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
```

**Argumentos**:

- options (Object):
    - host \(*String*\) **required**: identifica donde esta el server de redis, default: localhost
    - port \(*Number*\) **required**: puerto al cual se conecta redis, default: 6379
    - local \(*Boolean*\): activa el cache local, para mantener en memoria una copia del valor guardado, default: false

> Para mayor informacion sobre todos los campos aceptados, leer **[new redis][newRedis]**.

**Retorna**:

\(*ZIRedis*\): Retorna la instancia de la clase **ZIRedis**.

##### Método **nget**

Este método tiene su nombre por ser de nunchee y llamar al get de **[ioredis][ioredis]**. Se crea con el fin de mejorar la estrategia de get de ioredis, agregando sobre este método un redis local.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.set('foo', 'bar');
redis.nget('foo', ttlLocal)
    .then((bar) => /* bar => 'bar' */)
    .catch((err) => /* error interno de redis o ioredis */);
```

**Argumentos**:

- key \(*String*\) **required**: identifica el elemento que queremos obtener de redis
- ttlLocal \(*Number*\): tiempo para mantener el resultado en el cache local, por defecto **30 segundos**

**Retorna**:

- \(*String*\): Retorna el valor que este guardado en redis según la key pasada por parámetro


**Prueba de rendimiento**

```bash
$ yarn benchmark benchmark/nget-get.js

cache.get x 4,413 ops/sec ±34.20% (49 runs sampled)
cache.nget x 2,106,822 ops/sec ±8.20% (74 runs sampled)
cache.ttl x 7,762 ops/sec ±2.18% (74 runs sampled)
cache.exists x 7,782 ops/sec ±2.61% (69 runs sampled)
Fastest is cache.nget
```

Determinando que la implementación de **nget** supera notablemente **get** de **[ioredis][ioredis]**. Tambien se compara el rendimiento de **ttl** y **exists**


##### Método **nset**

Este método al igual que nget tiene su nombre por ser de nunchee y llamar al set de **[ioredis][ioredis]**. Se crea con el fin ayudar en la creación de las referencias a la información y poder poder utilizar la estrategia de redis local antes del redis de redis.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.nset('foo', 'bar');
redis.nset(key, value, seconds, secondsLocal);
```

**Argumentos**:

- key \(*String*\) **required**: identifica el elemento que queremos obtener de redis
- value \(*String*\) **required**: valor que se desea guardar, según la key pasada
- seconds \(*Number*\): segundos que se desea mantener en redis del valor, por defecto **60 segundos**
- secondsLocal \(*Number*\): segundos que se desea mantener local el valor, por defecto **30 segundos**


##### Método **ndel**

Este se crea al igual que nget y nset con el fin de mejorar la estrategia de get de ioredis.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.ndel('foo');
redis.ndel(key);
```

**Argumentos**:

- key \(*String*\) **required**: identifica el elemento que queremos eliminar de redis y del redis local.

##### Método **deletePattern**

Se combinan los métodos **[keys][redis-keys]** y **[del][redis-del]** de redis por simplicidad y comodidad.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.deletePattern('foo*')
    .then((totalDeleted) => {})
    .catch((err) => /* error interno de redis o ioredis */);
```

**Argumentos**:

pattern \(*String*\) **required**: patrón a seguir de los elementos a borrar.

**Retorna**:

\(*Number*\): Retorna la cantidad de elementos eliminados

##### Método **hmsetBulk**

Este metodo proviene del **[hmset][redis-hmset]** de redis solo que permite realizar el bulk del mismo.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.hmsetBulk({
    key1: {
        key1: 'value1',
        key2: 'value2',
        key3: 'value3',
        key4: 'value4',
    },
    key2: {
        key1: 'value1',
        key2: 'value2',
    },
    key3: {
        key1: 'value1',
        key2: 'value2',
        key3: 'value3',
        key4: 'value4',
        key5: 'value5',
    },
})
    .then((results) => {})
```

**Argumentos**:

elements \(*Object*\) **required**: objecto de objetos

**Retorna**:

\(*Array*\): Retorna el resultado de cada operación el cual puede ser error o el string `'ok'`

```javascript
// example
results = [
    [new ExampleError()],
    [null, 'OK'],
]
```

##### Método **hgetallBulk**

Este metodo proviene del **[hgetall][redis-hgetall]** de redis solo que permite realizar el bulk del mismo.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.hgetallBulk(['key1', 'key2', 'key3', 'key4'])
    .then((result) => {})
```

**Argumentos**:

keys \(*Array*\) **required**: listados de llaves a buscar

**Retorna**:

\(*Array*\): Listado de resultados por llave buscada la cual contiene 2 valores por llave buscada: el error si es que existe y el objecto con los valores guardados.

```javascript
// example
result = [
    [
        null, { // found
            key1: 'value1',
            key2: 'value2',
            key3: 'value3',
            key4: 'value4',
        },
    ],
    [
        null, { // found
            key1: 'value1',
            key2: 'value2',
        },
    ],
    [
        null, { // found
            key1: 'value1',
            key2: 'value2',
            key3: 'value3',
            key4: 'value4',
            key5: 'value5',
        },
    ],
    [
        null, { // not found
        },
    ],
]
```

##### Método **hfind**

Se combinan los métodos **[keys][redis-keys]** y `hgetallBulk` por simplicidad y comodidad.

```javascript
let redis = new ZIRedis({port: 6379, host: '127.0.0.1'});
redis.hfind('foo*')
    .then((results) => {})
    .catch((err) => /* error interno de redis o ioredis */);
```

**Argumentos**:

pattern \(*String*\) **required**: patrón a seguir de los elementos a traer.

**Retorna**:

\(*Object*\): Retorna el objecto con las keys encontradas o `null` en caso de no encontrar nada.

```javascript
// example
results = {
    key1: {
        key1: 'value1',
        key2: 'value2',
        key3: 'value3',
        key4: 'value4',
    },
    key2: {
        key1: 'value1',
        key2: 'value2',
    },
    key3: {
        key1: 'value1',
        key2: 'value2',
        key3: 'value3',
        key4: 'value4',
        key5: 'value5',
    },
}
```

## Scripts en **[lua][lua]**

Para agregar un nuevo script este debe ser creado como un archivo en la carpeta **[defineCommand][defineCommand]** con el nombre de la funcionalidad y con el siguiente formato de ejemplo:

```javascript
// ruta: defineCommand/deletePattern.js
'use strict';

module.exports = {
    name: 'deletePattern',
    definition: {
        numberOfKeys: 1,
        lua: 'return redis.call(\'del\', unpack(redis.call(\'keys\', KEYS[1])))',
    },
};
```

Ademas se tiene que agregar en el archivo **[index.js][defineCommandIndex]** el **require** al archivo del nuevo método:

```javascript
'use strict';

const newCommands = [
    require('./deletePattern.js'),
];

...
```

Para utilizar la nueva funcionalidad basta con instanciar la clase solamente.

```javascript
'use strict';

    redis = new ZIRedis();

redis.deletePattern('test*');
...
```

## Pruebas funcionales (Unit Testing)

Se llevaron a cabo las siguientes pruebas:

```bash
$ yarn test
```

### Pruebas de rendimiento (Benchmark)

Con el objetivo de mejorar la estrategia de redis del get de [ioredis][ioredis], se implemento nget y se compara su funcionalidad al ejecutar el siguiente comando:

```bash
$ yarn benchmark benchmark/nget-get.js

cache.get x 4,413 ops/sec ±34.20% (49 runs sampled)
cache.nget x 2,106,822 ops/sec ±8.20% (74 runs sampled)
cache.ttl x 7,762 ops/sec ±2.18% (74 runs sampled)
cache.exists x 7,782 ops/sec ±2.61% (69 runs sampled)
Fastest is cache.nget
```

Determinando que la implementación de **nget** supera notablemente **get** de **[ioredis][ioredis]**. Tambien se compara el rendimiento de **ttl** y **exists**

Otra prueba que se realizo en el contexto de implementación, es comparar una funcion **async/await** vs **promise**

```bash
$ yarn benchmark benchmark/nget-ngetAsync.js

cache.nget x 2,137,106 ops/sec ±7.18% (67 runs sampled)
ngetAsync x 1,807,064 ops/sec ±1.83% (78 runs sampled)
Fastest is cache.nget
```

Determinando que la implementación con **promise** es mas eficiente

Al comenzar a generar nuevos metodos con **bulk** de objectos se busco la forma más rapida de recorrerlos por lo cual se hizo el siguiente benchmark:

```bash
$ yarn benchmark benchmarck/object-elements.js

object.entries x 947,030 ops/sec ±1.52% (86 runs sampled)
object.keys x 4,141,516 ops/sec ±2.41% (82 runs sampled)
object for in x 4,715,356 ops/sec ±1.31% (83 runs sampled)
Fastest is object for in
```

Lo cual nos dejo como resultado que la forma mas rapida de recorrer un objecto es por medio de la sentencia **for...in**.

## Changelog

Todos los cambios importantes son escritos [aquí](CHANGELOG.md).

[defineCommand]: https://bitbucket.org/smartbox_way/nunchee-js/src/master/defineCommand
[defineCommandIndex]: https://bitbucket.org/smartbox_way/nunchee-js/src/master/defineCommand/index.js
[ioredis]: https://github.com/luin/ioredis
[newRedis]: https://github.com/luin/ioredis/blob/master/API.md#new_Redis
[redis]: https://redis.io
[redis-hgetall]: https://redis.io/commands/hgetall
[redis-hmset]: https://redis.io/commands/hmset
[redis-keys]: https://redis.io/commands/keys
[redis-del]: https://redis.io/commands/del
[lua]: https://redis.io/commands/eval
[istanbul]: https://istanbul.js.org/

CHANGELOG.md
## Changelog

Todos los cambios importantes son escritos aquí. El Formato esta basado en [Keep a Changelog](http://keepachangelog.com/es-ES/1.0.0/)

### [Unreleased]

### [1.1.0] - 2018-04-19
#### Changed
- Se modifica la opción de cache local, para que sea activada cuando se considere necesario

### [1.0.1] - 2018-01-07
#### Added
- Se utiliza [debug][debug], para visualizar la información relevante en el flujo de cache local y de redis
- Se agrega eslintignore

#### Changed
- Se modifica eslintrc para utilizar https://github.com/zerointermittency/eslint-config-zi

### [1.0.0] - 2018-01-07
#### Added
- Se agregan pruebas funcionales con el objetivo de tener probado todo el código, usando [istanbul js][istanbul] para saber cuanto
- Se agrego el metodo para buscar los hash por patrón (hfind)
- Se agrego el bulk para la operación hmset de redis (hmsetBulk)
- Se agrego el bulk para la operación hgetall de redis (hgetallBulk)
- Se implemento la funcionalidad para borrar **keys** de la DB con patrones (deletePattern)
- Se agrego una forma para añadir los scripts de **[lua][lua]** a las nuevas instancias de redis
- Se implemente en **nset** el tiempo de expiración
- Se agrega funcionalidad para setear (nset) y eliminar (ndel) información
- Se genera estrategia para el momento de obtener información, llamar al cache local, sobre el cache de redis
- README.md instalación, pruebas, uso y como contribuir al proyecto

[defineCommand]: https://bitbucket.org/smartbox_way/nunchee-js/src/master/defineCommand
[defineCommandIndex]: https://bitbucket.org/smartbox_way/nunchee-js/src/master/defineCommand/index.js
[ioredis]: https://github.com/luin/ioredis
[newRedis]: https://github.com/luin/ioredis/blob/master/API.md#new_Redis
[redis]: https://redis.io
[redis-hgetall]: https://redis.io/commands/hgetall
[redis-hmset]: https://redis.io/commands/hmset
[redis-keys]: https://redis.io/commands/keys
[redis-del]: https://redis.io/commands/del
[lua]: https://redis.io/commands/eval
[istanbul]: https://istanbul.js.org/
[debug]: https://www.npmjs.com/package/debug

test/index.js
'use strict';

const Mocha = require('mocha'),
    mocha = new Mocha({reporter: process.env.REPORTER || 'spec'}),
    Path = require('wrapper-path'),
    path = new Path(`${__dirname}/../`),
    chai = require('chai');

chai.use(require('chai-as-promised'));

global._path = path;
global._expect = require('chai').expect;
global._ZIRedis = require('../ZIRedis.js');

path.recursive
    .files('/test/', {exclude: /test\/index.js$/g})
    .forEach(file => mocha.addFile(file));

// Run the tests.
mocha.run(() => process.exit());


test/ZIRedis.js
'use strict';

describe('ZIRedis', () => {
    let cache = new _ZIRedis({port: 6379, host: '127.0.0.1', limit: 1, local: true}),
        cache1 = new _ZIRedis({port: 6379, host: '127.0.0.1', local: true}),
        cache2 = new _ZIRedis({port: 6379, host: '127.0.0.1'}),
        cache3 = new _ZIRedis({port: 6379, host: '127.0.0.1'});
    before(() => {
        cache.nset('asdf', 'qwerty');
    });
    it('success: instance ZIRedis', (done) => {
        let nc = new _ZIRedis({port: 6379, host: '127.0.0.1'});
        _expect(nc instanceof _ZIRedis).to.be.true;
        cache.set('foo', 'bar');
        cache.get('foo').then((bar) => {
            _expect(bar).to.be.equal('bar');
            done();
        }).catch(done);
    });
    it('catch: instance ZIRedis', () => {
        let nc1 = () => new _ZIRedis(6379, '127.0.0.1'),
            nc2 = () => new _ZIRedis('127.0.0.1:6379');
        _expect(nc1).to.throw('zi-redis: "opts" must be "object"');
        _expect(nc2).to.throw('zi-redis: "opts" must be "object"');
    });
    it('nget', (done) => {
        cache.nget('asdf').then((bar) => {
            _expect(bar).to.be.equal('qwerty');
            done();
        }).catch(done);
    });
    it('nset string', (done) => {
        cache.nset('qwerty', 'asdf');
        cache.nget('qwerty').then((value) => {
            _expect(value).to.be.equal('asdf');
            done();
        });
    });
    it('nset overwrite string', (done) => {
        cache.nset('qwerty', 'foo:bar');
        cache.nget('qwerty').then((value) => {
            _expect(value).to.be.equal('foo:bar');
            done();
        });
    });
    it('nset string with expire redis and local', function(done) {
        this.timeout(4000);
        cache.nset('qwerty', 'asdf', 1, 1);
        setTimeout(() => {
            cache.nget('qwerty').then((value) => {
                _expect(value).to.be.null;
                done();
            }).catch(done);
        }, 2000);
    });
    it('nset string with expire local', function(done) {
        this.timeout(4000);
        cache.nset('qwerty', 'asdf', 10, 1);
        setTimeout(() => {
            cache.nget('qwerty').then((value) => {
                _expect(value).to.be.equal('asdf');
                done();
            }).catch(done);
        }, 2000);
    });
    it('nset JSON.stringify and parse value', (done) => {
        cache.nset('a', JSON.stringify(['b', 'c', 'd']));
        cache.nget('a').then((value) => {
            _expect(JSON.parse(value)).to.be.an('array');
            done();
        }).catch(done);
    });
    it('catch nset: value must be a "string"', (done) => {
        const nset = () => cache.nset('a', ['b', 'c', 'd']);
        _expect(nset).to.throw('zi-redis: "value" must be "string"');
        done();
    });
    it('ndel', (done) => {
        cache.nset('qwerty', 'asdf');
        cache.ndel('qwerty');
        cache.nget('qwerty').then((value) => {
            _expect(value).to.be.null;
            done();
        }).catch(done);
    });
    it('multpiples instances of cache', function(done) {
        this.timeout(50000);
        cache1.ndel('foo1');
        cache2.nget('foo1', 1)
            .then((value) => {
                _expect(value).to.be.null;
                return Promise.resolve();
            })
            .then(() => cache3.nget('foo1', 1))
            .then((value) => {
                cache1.nset('foo1', 'bar1', 20, 20);
                cache1.nset('foo2', 'bar2', false);
                _expect(value).to.be.null;
                setTimeout(() => Promise.resolve(), 3000);
            })
            .then(() => cache2.nget('foo1'))
            .then((value) => {
                _expect(value).to.be.equal('bar1');
                return Promise.resolve();
            })
            .then(() => cache2.nget('foo2'))
            .then((value) => {
                _expect(value).to.be.equal('bar2');
                return Promise.resolve();
            })
            .then(() => cache3.nget('foo1', 1))
            .then((value) => {
                _expect(value).to.be.equal('bar1');
                return Promise.resolve();
            })
            .then(() => cache1.nget('foo1', 1))
            .then((value) => {
                _expect(value).to.be.equal('bar1');
                return Promise.resolve();
            })
            .then(() => cache3.nget('foo1', 1))
            .then(value => new Promise((res) => {
                _expect(value).to.be.equal('bar1');
                setTimeout(() => res(), 3000);
            }))
            .then(() => done())
            .catch(done);
    });
});


test/.eslintrc.js
'use strict';

module.exports = {
    globals: {
        _path: true,
        _expect: true,
        _ZIRedis: true,
    },
};


benchmark/nget-get.js
'use strict';

const ZIRedis = require('../ZIRedis.js'),
    cache = new ZIRedis({
        host: 'localhost',
        port: 6379,
        db: 0,
        readOnly: true,
        local: true,
    }),
    Benchmark = require('benchmark');

cache.set('foo', 'bar');

/* eslint-disable no-console */
(new Benchmark.Suite)
    .add('cache.get', {
        defer: true,
        fn: deferred => cache.get('foo').then(() => deferred.resolve()),
    })
    .add('cache.nget', {
        defer: true,
        fn: deferred => cache.nget('foo').then(() => deferred.resolve()),
    })
    .add('cache.ttl', {
        defer: true,
        fn: deferred => cache.ttl('foo').then(() => deferred.resolve()),
    })
    .add('cache.exists', {
        defer: true,
        fn: deferred => cache.exists('foo').then(() => deferred.resolve()),
    })
    .on('cycle', event => console.log(String(event.target)))
    .on('complete', function() {
        console.log('Fastest is ' + this.filter('fastest').map('name'));
        process.exit();
    })
    .run({'async': false});


benchmark/eventLoop.js
'use strict';

const EventEmitter = require('events'),
    reader = new require('ioredis')({
        connectionName: 'opta-reader',
        host: 'localhost',
        port: 6379,
        db: 0,
        dropBufferSupport: true,
        enableReadyCheck: true,
        readOnly: true,
    });

class Client extends EventEmitter {
    eventTest(index, done) {
        this.emit('event', {index, done});
    }
}

const anotherCallbackFunction = (index, cb) => {
    console.log(`test ${index} -- 4`);
    (() => {
        console.log(`test ${index} -- 5`);
        cb();
    })();
};

const callbackFunction = (index, cb) => {
    console.log(`test ${index} -- 2`);
    (() => {
        setTimeout(() => {
            console.log(`test ${index} -- 3`);
            anotherCallbackFunction(index, cb);
        }, 10);
    })();
};

// const anotherCallbackFunction = (index, cb) => {
//     console.log(`test ${index} -- 4`);
//     reader.get(`${index}`, () => {
//         console.log(`test ${index} -- 5`);
//         cb();
//     });
// };

// const callbackFunction = (index, cb) => {
//     console.log(`test ${index} -- 2`);
//     reader.get(`${index}`, () => {
//         console.log(`test ${index} -- 3`);
//         anotherCallbackFunction(index, cb);
//     });
// };

const client = new Client();

// client.on('event', (index) => {
//     console.log(`test ${index} -- 1`);
//     callbackFunction(index, () => {
//         console.log(`test ${index} -- 6`);
//     });
// });

// for (let i = 0; i < 20; i++) {
//     client.eventTest(i);
// }

client.on('event', ({index, done}) => {
    console.log(`test ${index} -- 1`);
    callbackFunction(index, () => {
        console.log(`test ${index} -- 6`);
        done(null);
    });
});

const each = require('async/each'),
    listEvents = [];

for (let i = 0; i < 1000000; i++)
    // if (listEvents.length == 0)
    listEvents.push(i);
    // else
    //     listEvents.push((previus, done) => client.eventTest(i, done));

each(listEvents, (i, done) => client.eventTest(i, done), (err, result) => {
    console.log(err);
    console.log(result);
    console.log('end all events');
    process.exit();
});


benchmark/nget-ngetAsync.js
'use strict';

const ZIRedis = require('../ZIRedis.js'),
    cache = new ZIRedis({
        host: 'localhost',
        port: 6379,
        db: 0,
        readOnly: true,
    }),
    cache1 = new ZIRedis({
        host: 'localhost',
        port: 6379,
        db: 0,
        readOnly: true,
    }),
    cache2 = new ZIRedis({
        host: 'localhost',
        port: 6379,
        db: 0,
        readOnly: true,
    }),
    Benchmark = require('benchmark');

cache.set('foo', 'bar');

async function ngetAsync(key, local = 10) {
    const self = cache2;
    if (!self._keys[key]) {
        const value = await self.get(key);
        if (!value) return value;
        let ttl = await self.ttl(key);
        self._keys[key] = key;
        if (ttl === -1) ttl = local;
        self._cache.set(self._keys[key], Promise.resolve(value));
        setTimeout(() => {
            self._cache.delete(self._keys[key]);
            delete self._keys[key];
        }, ttl * 1000);
        return value;
    }
    return self._cache.get(self._keys[key]);
}

// cache.nget('foo').then((value) => {
//     console.log('#value', require('util').inspect(value, 0, 10, 1));
// });

// ngetAsync('foo').then((value) => {
//     console.log('#value', require('util').inspect(value, 0, 10, 1));
// });

/*eslint-disable no-console */
(new Benchmark.Suite)
    .add('cache.nget', {
        defer: true,
        fn: (deferred) => cache1.nget('foo').then(() => deferred.resolve()),
    })
    .add('ngetAsync', {
        defer: true,
        fn: (deferred) => ngetAsync('foo').then(() => deferred.resolve()),
    })
    .on('cycle', (event) => console.log(String(event.target)))
    .on('complete', function() {
        console.log('Fastest is ' + this.filter('fastest').map('name'));
        process.exit();
    })
    .run({'async': false});

benchmark/object-elements.js
'use strict';

const Benchmark = require('benchmark'),
    object = {
        'test1': 'test1',
        'test2': 'test2',
        'test3': 'test3',
        'test4': 'test4',
        'test5': 'test5',
        'test6': 'test6',
        'test7': 'test7',
        'test8': 'test8',
        'test9': 'test9',
        'test10': 'test10',
    };

/* eslint-disable no-console */
(new Benchmark.Suite)
    .add('object.entries', () => {
        const values = Object.entries(object);
        for (let i = values.length - 1; i >= 0; i--) {
            values[i][0];
            values[i][1];
        }
    })
    .add('object.keys', () => {
        const keys = Object.keys(object);
        for (let i = keys.length - 1; i >= 0; i--) {
            const key = keys[i];
            object[key];
        }
    })
    .add('object for...in', () => {
        for (let key in object)
            object[key];
    })
    .on('cycle', event => console.log(String(event.target)))
    .on('complete', function() {
        console.log('Fastest is ' + this.filter('fastest').map('name'));
        process.exit();
    })
    .run({'async': false});


ZIRedis.js
'use strict';

const Redis = require('ioredis'),
    defineCommands = require('./defineCommand'),
    core = {
        errors: require('./core/errors.js'),
        debug: require('./core/debug.js'),
    },
    ttlRedis = 60,
    ttlLocal = 30,
    timeout = (debug, self, key, ttl) => {
        setTimeout(() => {
            debug('delete local', key);
            self._cache.delete(self._keys[key]);
            delete self._keys[key];
        }, ttl * 1000);
    };

class ZIRedis extends Redis {
    constructor(opts) {
        const debug = core.debug('constructor');
        debug('opts', require('util').inspect(opts, 0, 10, 1));
        if (!opts || typeof opts !== 'object' || Array.isArray(opts))
            throw core.errors.type('opts', 'object');
        super(opts);

        if (opts.local) {
            this._keys = {};
            this._limit = opts.limit || 500;
            this._cache = new Map();
        }

        defineCommands(this);
    }

    nset(key, value, seconds = ttlRedis, local = ttlLocal) {
        const self = this,
            debug = core.debug('nset');
        debug('key', key);
        debug('value', value);
        debug('ttlRedis', seconds);
        debug('ttlLocal', local);
        if (typeof value !== 'string')
            throw core.errors.type('value', 'string');
        /* istanbul ignore else */
        if (self._cache) {
            if (Object.keys(self._keys).length === self._limit) {
                self._keys = {};
                self._cache.clear();
            }
            /* istanbul ignore else */
            if (!self._keys[key]) self._keys[key] = key;
            self._cache.set(self._keys[key], Promise.resolve(value));
            timeout(debug, self, key, local);
        }
        if (seconds) return self.set(key, value, 'EX', seconds);
        self.set(key, value);
    }

    ndel(key) {
        const self = this,
            debug = core.debug('ndel');
        debug('key', key);
        if (self._cache && self._keys[key]) {
            self._cache.delete(self._keys[key]);
            delete self._keys[key];
        }
        return self.del(key);
    }

    nget(key, local) {
        const self = this,
            debug = core.debug('nget');
        debug('key', key);
        debug('local', local);
        if (self._cache) {
            if (!self._keys[key])
                return new Promise((res, rej) => {
                    let result;
                    self.get(key)
                        .then((value) => {
                            if (!value) return res(value);
                            result = value;
                            return super.ttl(key);
                        })
                        .then((ttl) => {
                            if (!ttl) return;
                            self._keys[key] = key;
                            /* istanbul ignore next */
                            if (ttl === -1) ttl = ttlLocal;
                            /* istanbul ignore next */
                            if (local) ttl = local;
                            self._cache.set(self._keys[key], Promise.resolve(result));
                            timeout(debug, self, key, ttl);
                            debug('result', result);
                            res(result);
                        })
                        .catch(rej);
                });

            return self._cache.get(self._keys[key])
                .then((value) => {
                    debug('result', value);
                    return value;
                });
        } else
            return self.get(key);
    }

    deletePattern(pattern) {
        const self = this,
            debug = core.debug('deletePattern');
        debug('pattern', pattern);
        return new Promise((resolve, reject) => {
            if (typeof pattern !== 'string')
                return reject(core.errors.type('pattern', 'string'));
            if (pattern.length === 0)
                return reject(core.errors.empty('pattern'));

            self.keys(pattern)
                .then((keys) => {
                    if (keys.length === 0)
                        return 0;
                    return self.del(keys);
                })
                .then(resolve)
                .catch(reject);
        });
    }

    hmsetBulk(elements) {
        const self = this,
            debug = core.debug('hmsetBulk');
        debug('elements', elements);
        return new Promise((resolve, reject) => {
            if (!elements || typeof elements !== 'object' || Array.isArray(elements))
                return reject(core.errors.type('elements', 'object'));

            const pipe = self.pipeline();

            let hasElements = false;
            for (let key in elements) {
                hasElements = true;
                pipe.hmset(key, elements[key]);
            }

            if (hasElements === false)
                return reject(core.errors.empty('elements'));

            pipe.exec()
                .then(resolve)
                .catch(reject);
        });
    }

    hgetallBulk(keys) {
        const self = this,
            debug = core.debug('hgetallBulk');
        debug('keys', keys);
        return new Promise((resolve, reject) => {
            if (!Array.isArray(keys))
                return reject(core.errors.type('keys', 'array'));
            if (keys.length === 0)
                return reject(core.errors.empty('keys'));

            const pipe = self.pipeline();

            for (let i = 0; i < keys.length; i++)
                pipe.hgetall(keys[i]);

            pipe.exec()
                .then(resolve)
                .catch(reject);
        });
    }

    hfind(pattern) {
        const self = this,
            debug = core.debug('hfind');
        debug('pattern', pattern);
        return new Promise((resolve, reject) => {
            if (typeof pattern !== 'string')
                return reject(core.errors.type('pattern', 'string'));
            if (pattern.length === 0)
                return reject(core.errors.empty('pattern'));

            let keys;
            self.keys(pattern)
                .then((_keys) => {
                    keys = _keys;
                    if (keys.length === 0) return keys;
                    return self.hgetallBulk(_keys);
                })
                .then((values) => {
                    if (values.length === 0) return resolve(null);
                    const result = {};
                    for (let i = 0; i < keys.length; i++)
                        result[keys[i]] = values[i][1];
                    resolve(result);
                })
                .catch(reject);
        });
    }
}

module.exports = ZIRedis;
